README
###########################################################################
# PM_tool                                                                 #
# version 1.2.3                                                           #
#                                                                         #
# Copyright (C) Robby Russell 2004 - 2005           		          #
# email: robby@planetargon.com 	                              	          #
# www  : http://www.pmtool.org/					          #
#                                                                         #
# This program is free software; you can redistribute it and/or modify    #
# it under the terms of the GNU General Public License as published by    #
# the Free Software Foundation; either version 2 of the License, or       #
# (at your option) any later version.                                     #
#                                                                         #
# This program is distributed in the hope that it will be useful,         #
# but WITHOUT ANY WARRANTY; without even the implied warranty of          #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           #
# GNU General Public License for more details.                            #
#                                                                         #
# You should have received a copy of the GNU General Public License       #
# along with this program; if not, write to the Free Software             #
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.               #
#
###########################################################################
# Original Author: Olaf Willuhn 2000 - 2002 				  #
###########################################################################

###########################################################################
# NOTE: We promise to make this much easier in the near future. :-)
###########################################################################

###########################################################################
# REQUIREMENTS                                                            #
###########################################################################

- Apache Webserver 1.3.6+ (or any other webserver, that supports PHP)
  (http://www.apache.org/)

- MySQL Server 3.22+
  (http://www.mysql.com/)

- PHP 4+ compiled with MySQL-Support
  (http://www.php.net/)

###########################################################################
# UPDATE                                                                  #
###########################################################################

If you upgrade from a previous version (0.9.x or 1.0.x) please follow these
steps instead of the notes in the "INSTALLATION" section

- just copy all the files into any dir in your webserver documents dir
  (i.e.: /usr/local/apache/htdocs/foo/bar/pmtool_1.0)
  You can also delete all old pmtool files and copy the new ones
  into this dir. (of course, you should make a backup of the old install ;)

- edit your new configfile ("cfg/config.inc.php")

  set the following vars to your settings
  $dbconfig['user'] = "<pmtool_user>";     // db username
  $dbconfig['pass'] = "<pmtool_password>"; // db password
  $dbconfig['host'] = "<hostname>";        // hostname[:port]
  $dbconfig['name'] = "<databasename>";    // databasename

- if you update from version 0.9.x:
  Run the SQL-Update-Script:
  $> mysql -u <pmtool_user> -p<pmtool_password> < sql/table_update_0.9.x.sql

- if you update from version 1.0.x:
  Run the SQL-Update-Script:
  $> mysql -u <pmtool_user> -p<pmtool_password> < sql/table_update_1.0.x.sql

  Note: in 1.1 there are some new permission checks for task attachment.
        If you uprade from 1.0.x, you need to add the following rights:
          - attachment.activate (groups: admin,developer,manager,customer)
          - attachment.insert   (groups: admin,developer,manager,customer)
          - attachment.delete   (groups: admin,developer,manager)
          - attachment.getSize  (groups: admin,developer,manager,customer)
          - task.isAvailable    (groups: admin,developer,manager,customer)
          - task.fixedPrice     (groups: admin,developer,manager,customer)

        manually on the "rights" page of pmtool.
        This is not needed, if you upgrade from 0.9.x or if you install
        pmtool "from scratch"

- if you update from version 1.1.x (before 1.1e):
  Run the SQL-Update-Script:
  $> mysql -u <pmtool_user> -p<pmtool_password> < sql/table_update_1.1.x.sql

- if you update from version 1.2:
  Run the SQL-Update-Script:
  $> mysql -u <pmtool_user> -p<pmtool_password> < sql/table_update_1.2.x.sql

- if you update from version 1.2.1:
  Run the SQL-Update-Script:
  $> mysql -u <pmtool_user> -p<pmtool_password> < sql/table_update_1.2.1.sql

- if you update from version 1.2.2:
  Run the SQL-Update-Script:
  $> mysql -u <pmtool_user> -p<pmtool_password> < sql/table_update_1.2.3.sql

###########################################################################
# INSTALLATION                                                            #
###########################################################################

- just copy all the files into any dir in your webserver documents dir
  (i.e.: /usr/local/apache/htdocs/foo/bar/pmtool)

- create a directory "run" writable for the webserver!
  This is needed, if you want to upload task attachments
  $> mkdir -p run/attach
  $> chmod -R 757 run

- create a MySQL-database and MySQL-User for pmtool
  <pmtool_user>     -> database username
  <pmtool_password> -> database password

  $> mysql -u root -p
  <enter your password here>

  Welcome to the MySQL monitor.  Commands end with ; or \g.
  Your MySQL connection id is 5525 to server version: 3.22.32

  Type 'help' for help.

  mysql> create database pmtool;
  mysql> grant usage on pmtool.* to <pmtool_user>@localhost;
  mysql> grant select,insert,update,delete,create on pmtool.* to <pmtool_user>@localhost identified by '<pmtool_password>';
  mysql> flush privileges;
  mysql> exit

- create MySQL-Tables
  $> mysql -u <pmtool_user> -p<pmtool_password> < sql/table_create.sql

- edit the configfile ("cfg/config.inc.php")

  set the following vars
  $dbconfig['user'] = "<pmtool_user>";     // db username
  $dbconfig['pass'] = "<pmtool_password>"; // db password
  $dbconfig['host'] = "<hostname>";        // hostname[:port]
  $dbconfig['name'] = "<databasename>";    // databasename

  don't forget "<ESC>:wq" to store the changes in your config ;-)

###########################################################################
#                                                                         #

- point your favorite webbrowser to "http://foo.bar/foo/bar/pmtool/index.php",
  login as user "admin" (without password), go to "users", give yourself
  a password. After that, go to "groups" and put your users into the different
  groups.
  NOTE: the group "access" includes all needed rights for changing the ACLs.
        You should put a kind of backdoor-user into this group.
        Reason: It's possible to remove yourself from ALL groups/rights!
        After this, you will not be able to login!
        !!! So, be VERY carefully, if you change the rights and groups !!!

- COOL GIMMICK:
  if you have installed windows on your client PC, you can add the
  pmtool taskbar to your start menu. This is a very usefull small bar
  to hanlde your jobs. So you do not need to open a browser window every time
  you post your jobs. You can see a screenshot of this at
  http://www.pmtool.org/index.php?page=5

  INSTALL: - right-click on your start menu
           - choose "toolbars" (i hope, this is right, it's named
             "Symbolleisten" in german ;)
           - select the entry "new toolbar..."
           - windows will now prompt you to enter a folder name
           - instead of this, you can also enter URLs starting with "http*"
           - so type "http://<your.pmtool.location>/taskbar.php" in the textfield
           - press the "OK" button
           - pretty cool, huh ? ;)

Note : I know, there's no documentation for this thing, just
       experience it - it's pmtool XP ;)

PS   : sorry for my sucking english, but i'm german :o)

PPS  : NEVER, NEVER use Netscape 4.x as Browser for pmtool - it looks
       really UGLY! ;)

PPPS : pmtool contains a feature to generate reports written in XML and
       designed by XSL-templates. Currently only MS Internet Explorer
       is able to show such files!
       You can create own XSL-Templates like the example in
       templates/reports/bill/template.xsl. Just create a new
       file "template/reports/<foo bar>/template.xsl
       it will automatically be shown in the report dialog

#                                                                         #
###########################################################################



ChangeLog
2005-02-25 Robby Russell
  @N added new Portuguese/Brazilian translation.
  @N removed PHPSESSID from urls for security purposes.
  @N Cleaned up inc/includer.inc.php
  @N Cleaned up index.php
  @N Fixed bugs with times displaying as 0:00 in the task viewer window.
  @N Moved constants
  @N Began cleaning up inc/users.inc.php for future db layer

2004-03-17 Olaf Willuhn <info@willuhn.de>
  @N added the ability to configure the behavior for which kind of tasks the
     customer has to pay for. (see cfg/config.inc.php --> $config['paytasks'])
  @N added project priorities.
     NOTE: you have to apply the latest update sql script (table_update_1.2.1.sql to add this row)
     --> now we have two classes which have an attribute "priority", so i've
         moved the according code into a separate class "priority.inc.php".
         both classes (task (via request) and project) inherit from this one.

2004-02-29  <>
  @B missing semicolon at the and of the line

2004-02-29  <>
  @N added tom's new column "flags" to the job table

2004-02-17  <>
  @N added consecutive id to xml code of reports
  @N added czech language file
  @N added spanish language file

2003-11-18  <>
  @N updated italian language file

2003-11-17  <>
  @N some more fixes at the new project status plugin

2003-11-05  <>
  @N added portuguese language file  - many thanks to Daniel Santos
  @C changed the field "fied price" in tasks from integer to float

2003-10-09 Olaf Willuhn <info@willuhn.de>
  @B bug fixed in encodeString and rights page (array were not handled correctly via POST)

2003-10-07 Olaf Willuhn <info@willuhn.de>
  @B fixed bug (sourceforge bug id: 815180)

2003-10-07 Olaf Willuhn <info@willuhn.de>
  @N added italian language file

2003-09-27 Olaf Willuhn <info@willuhn.de>
  @B fix on logon page (empty passwords enabled again)
  @N added DEMO_MODE

2003-09-27 Olaf Willuhn <info@willuhn.de>
  @B fixed typo in taskbar.php

2003-09-27 Olaf Willuhn <info@willuhn.de>
  @C replaced all HTTP_POST_VARS and HTTP_GET_VARS with tool::secureGet() and tool::securePost()
  @C replacing of all dangerous chars in input (cross site scripting)

2003-09-02 Olaf Willuhn <info@willuhn.de>
  @N added dutch language file - many thanks to Suat Gedik

2003-08-29 Olaf Willuhn <info@willuhn.de>
  @C changed xml column in report table from blob to mediumblob

2003-08-27: v1.1b
      + new   : aded some sql indexes (for performance reasons)
      + new   : tasks without jobs but fixed price can be reported now

2003-08-08:
      + bug   : see bug id 771449 on sf.net

2003-07-27: v1.1a
      + change: removed all "short open tags"
                "&lt;?" --> "&lt;?PHP"
                --==[ Many Thanks to Arne Kepp who made this! ]==--

2003-01-10: v1.1
      + new   : attachments for tasks
      + new   : task dependencies
      + new   : support for plugins (some sample plugins included)
      + new   : multilanguage support (incomplete)
      + new   : some new fields in tasks
      + new   : bugfixes in session handling, a lot more ;)

2001-12-11: v1.0.1
      + new   : some new tags in reports
                see templates/report/bill/syntax.xml
                for details
      + bug   : security bugfix in index.php (include())
      + new   : rounding of used time implemented

2001-12-05: v1.0
      + new   : completly redesigned the whole application
                encapsulated all functions in classes
      + new   : ability to generate reports in XML format
      + new   : ACLs
                you can create/change groups, put users
                into and assign rights to the groups
                You are able to control the access for
                every function call!
      + new   : customer logins
      + new   : ability to set costs per hour for users
                AND projects. if both ist set, the project
                costs will be used.
      + new   : requests (most usefull for customers)
                users are able to post requests. the
                project manager will be informed by email
                and is able to assign this request to one
                of his developers. the request will then
                switched into a task
      + laberlaber ... foobar ... many other things ;)

2001-08-21: v0.9.6
      + new   : new design
      + new   : db functions in class style

2001-08-15: v0.9.5
      + new   : only users with admin access are now allowed
                to change user records.
                All other users may only change their own password

        If you are upgrading from a previous version, please
        execute the following commands at mysql prompt:

        mysql> use pmtool;
        mysql> update user set access='a' where username='<your username>';
        mysql> exit;

        This is needed, to give yourself initial admin access.

2001-08-05: v0.9.4a
      + bugfix: default start time on report page
                was calculated wrong

2001-07-28: v0.9.4
      + bugfix: session_start in index.php
      + bugfix: forward url in logout.php
        Thanks to Thomas Kho again ;)
      + new   : new page "reports" to create a list
                of your jobs for the last days. Good
                for creating status reports for your
                project manager ;)

2001-07-15: V0.9.3
      + new   : support for creating excel-sheets on the
                fly (in stats page) requires the PERL-Modul
                "Spreadsheet::WriteExcel" !
                See README for further details
      + new   : you can sort all pages by clicking the
                table header, a click again reverts the order
      + new   : taskbar utility
                this is a VERY usefull tool for windows users
                - just rightclick on your startmenu, choose
                  iconbar, new iconbar... and place the url
                  "http://<pmtool-location>/taskbar.php"
                  into the field "foldername" - it seems to be

taskbar.php
<?PHP
/***************************************************************************
 * $Source: /cvsroot/pmtool/pmtool/taskbar.php,v $
 * $Revision: 1.6 $
 * $Date: 2003/12/07 19:21:16 $
 * $Author: willuhn $
 * $Locker:  $
 * $State: Exp $
 *
 * Copyright (c) by willuhn.webdesign
 * All rights reserved
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 ***************************************************************************/

if (!file_exists("inc/includer.inc.php")) {echo "panic: inc/includer.inc.php doesn't exist";exit;} require("inc/includer.inc.php");

if (tool::securePost('action') && tool::securePost('action') == "logout") {
  session_start();
  session_destroy();
//  header("Location: ".$HTTP_SERVER_VARS['PHP_SELF']);
}


?>
<html>
<head>
  <title>PMtool</title>
  <meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=<?php echo $lang['charset']?>" />
  <link rel=stylesheet type="text/css" href="styles.css" />
  <script language="JavaScript" src="js/common.js" type="text/javascript"></script>
</head>
<body class="taskbar">
<div>
<form method="post" name="form1" class="taskbar" target="_self">
<?PHP

if (! isset($HTTP_SESSION_VARS["loginid"]) || $HTTP_SESSION_VARS["loginid"] == "") {
  // session is not set -> authenticate

  // try to authenticate by IP
  if ($loginInst->authByIp()) {
    $HTTP_SESSION_VARS['loginid'] = $loginInst->authByIp();
  }
  // try to authenticate by username/password
  elseif (tool::securePost('loginname') && tool::securePost('password') && $loginInst->authByPassword(tool::securePost('loginname'),tool::securePost('password'))) {
    $HTTP_SESSION_VARS['loginid'] = $loginInst->authByPassword(tool::securePost('loginname'),tool::securePost('password'));
  }

  if (isset($HTTP_SESSION_VARS['loginid']) && $HTTP_SESSION_VARS['loginid'] != "" && ! session_is_registered("loginid")) {
    $loginid = $HTTP_SESSION_VARS['loginid'];
    if (! session_register("loginid")) {
      echo "<b>".$lang['common_unableToSaveLoginInSession']."</b><br>";
      // could not save session -> give up
      exit;
    }
  }

  elseif (tool::securePost('loginname') || tool::securePost('password')) {
    // show error message only, if username/password was submitted
    $toolInst->errorStatus($lang['common_userUnknownOrPasswordWrong']);
  }
}


///// Language settings
$lang = array();
$importLang = "en";
$langFound = 0;

// check if language default is set
if (file_exists("lang/".$config['language'].".inc.php")) {
  $langFound = 1;
}


if (session_is_registered("loginid")) {
  if (! isset($HTTP_SESSION_VARS['loginid']) || $HTTP_SESSION_VARS['loginid'] == "") {
    echo "<b>".$lang['common_unableToFindloginInSession']."</b><br>";
    // could not save session -> give up
    exit;
  }

  // activate user
  $loginInst->activate($HTTP_SESSION_VARS['loginid']);

  // check existence of user specific language
  if (isset($loginInst->language) &&
      $loginInst->language != "" &&
      file_exists("lang/".$loginInst->language.".inc.php"))
  {
    $importLang = $loginInst->language;
    $langFound = 1;
  }

  if ($langFound == 1) {
    require("lang/".$importLang.".inc.php");
  }
  else {
    echo "panic, no language file found.\n";
    exit;
  }

}
else {

  if ($langFound == 1) {
    require("lang/".$importLang.".inc.php");
  }
  else {
    echo "panic, no language file found.\n";
    exit;
  }


  ?>
  <table border="0" cellpadding="0" cellspacing="0" width="100%">
    <tr>
      <td>
        <b><?PHP echo $lang['common_login'];?>:</b>
        <?PHP echo $lang['common_name'];?>:
        <?PHP
          $selected = "";
          if (tool::securePost('loginname') && tool::securePost('loginname') != "") {
            $selected = tool::securePost('loginname');
          }
          $dbInst->selectbox($dbInst->config['table_user'],"name","username","loginname",$selected);
        ?>
        &nbsp;<?PHP echo $lang['common_password'];?>: <input type="password" name="password" size="<?PHP echo $htmlconfig['text_size3'];?>">
        <input type="submit" value="<?PHP echo $lang['common_login'];?>">
      </td>
      <td align="right">
        <?PHP if (isset($global_status)) {
          echo $global_status;
        }?>
      </td>
    </tr>
  </table>
  </form>
  </body>
  </html>
  <?PHP
  exit;
}


if (tool::securePost('action')) {
  if (tool::securePost('action') == "start") {
    $jobInst = new job();
    $jobInst->taskId = tool::securePost('taskid');
    $jobInst->userId = $loginInst->id;
    $jobInst->comment = tool::securePost('comment');
    $jobInst->start();

  }
  elseif (tool::securePost('action') == "stop") {
    $jobInst = new job(tool::securePost('jobid'));
    $jobInst->comment = tool::securePost('comment');
    $jobInst->stop();
    if (tool::securePost('taskdone') && tool::securePost('taskdone') == "1") {
      $taskInst = new task($jobInst->taskId);
      $taskInst->stop();
      $taskInst->update();
    }
  }
}

?>
<table border="0" cellpadding="0" cellspacing="0" width="100%">
  <tr>
    <td>
      <?PHP

      $jobInst = new job();

      // check, if there's an open job
      $jobId = $jobInst->getOpenJob();
      if (isset($jobId) && $toolInst->checkInt($jobId)) {
        $jobInst->activate($jobId);
        $taskInst = new task($jobInst->taskId);
        $projectInst = new project($taskInst->projectId);
        ?>
        <input type="hidden" name="jobid" value="<?PHP echo $jobInst->id;?>">
        <input type="hidden" name="taskid" value="<?PHP echo $jobInst->taskId;?>">
        <input type="hidden" name="action" value="stop">
        <a href="javascript:document.form1.action.value='logout';document.form1.submit()" target="_self"><?PHP echo $lang['common_logout'];?></a>
        &nbsp;[<a target="_self" title = "<?PHP echo $lang['common_klickToRefreshThisTimeRunningSince'];?> <?PHP echo $toolInst->getTime("Y-m-d, H:i",$jobInst->start);?>)" href="<?PHP echo $toolInst->encodeUrl("taskbar.php");?>">time</a>: <input readonly="" style="width:54px;font-size:8pt;font-family:Verdana, Helvetica, Arial;text-align:right;border:0;background:transparent" size="5" name="servertime" value="<?PHP echo $toolInst->formatTime($jobInst->getSummary());?>">]&nbsp;<b><a target="_new" title="klick to view tasklist in browser" href="<?PHP echo $toolInst->encodeUrl("index.php?content=tasks.php");?>">task</a>:</b>&nbsp;<span title="<?PHP echo htmlspecialchars($taskInst->body);?>" onclick="javascript:openwindow('<?PHP echo $toolInst->encodeUrl("index.php?content=taskdetails.php&view=details&taskid=".$taskInst->id);?>',width='500',height='500')"><?PHP echo htmlspecialchars(substr("[".$taskInst->id."] ".$projectInst->name.": ".$taskInst->subject,0,28))?></span>&nbsp;
        <b><?PHP echo $lang['common_comment'];?>:</b>&nbsp;<input value="<?PHP echo $jobInst->comment;?>" type="text" name="comment" size="<?PHP echo $htmlconfig['text_size3'];?>"><input title="<?PHP echo $lang['common_closeTask'];?>" type="checkbox" class="taskbar" name="taskdone" value="1"><input type="submit" value="<?PHP echo $lang['common_stop'];?>">
        <script language="JavaScript" type="text/javascript">
          <!--//
          <?PHP
            $min_counter = explode(":",preg_replace("/ h/","",$toolInst->formatTime($jobInst->getSummary())));
          ?>
          var sec = 0;
          var minute='<?PHP echo $min_counter[1];?>';
          var hour='<?PHP echo $min_counter[0];?>';

          function calctime(sec2,minute2,hour2){
            if (sec2>59) {
              sec2-=60;
              minute2++;
            }
            if (minute2>59) {
              minute2-=60;
              hour2++;
            }
            if (hour2>23) {
              hour2-=24;
            }
            var nowtime = new Date();
            var d="";
            if(hour2<10) d="0";
            d += hour2+":";
            if(minute2<10) d+="0";
            d+=minute2+":";
            if(sec2<10) d+="0";
            d+=sec2;

            form1.servertime.value=d;
            setTimeout('calctime('+(sec2+1)+','+minute2+','+hour2+')',1000);
            return true;
          }
          calctime(sec+1,minute,hour);
          //-->
        </script>
        <?PHP
      }

      // else: show task list
      else {
        ?>
        <input type="hidden" name="action" value="start">
        <a href="javascript:document.form1.action.value='logout';document.form1.submit()" target="_self"><?PHP echo $lang['common_logout'];?></a>
        &nbsp;[<a target="_self" title="<?PHP echo $lang['common_klickToRefreshTasklist'];?>" href="<?PHP echo $toolInst->encodeUrl("taskbar.php");?>"><?PHP echo $lang['common_reload'];?></a>]
        <b><a title="<?PHP echo ;?>" href="<?PHP echo $toolInst->encodeUrl("index.php?content=tasks.php");?>" target="_new"><?PHP echo $lang['common_task'];?></a>:</b>&nbsp;
        <?PHP
        $taskInst = new task();
        $taskInst->filterUserId = $loginInst->id;
        $taskInst->filterStatusId = TASK_STATUS_DONE;
        $taskInst->filterInvertStatus = "1";
        $list = $taskInst->getList("project");
        if (! $list) {
          echo $lang['common_noOpenTasksAvailable'];
        }
        else {
          echo "<select name=\"taskid\">\n";
          while ($element = current($list)) {
            $taskInst->activate($element);
            if ($taskInst->isAvailable()) {
              $projectInst = new project($taskInst->projectId);
              $selected = "";
              if (tool::securePost('taskid') && $taskInst->id == tool::securePost('taskid')) $selected = "selected";
              echo "<option ".$selected." value=".$taskInst->id.">[".$taskInst->id."] ".substr($projectInst->name.":&nbsp;".$taskInst->subject,0,55)."...\n";
            }
            next($list);
          }
          echo "</select>&nbsp;\n";
          ?>
          <b><?PHP echo $lang['common_comment'];?>:</b>&nbsp;
          <input type="text" name="comment" size="<?PHP echo $htmlconfig['text_size3'];?>"><input type="submit" value="<?PHP echo $lang['common_start'];?>">
        <?PHP } ?>
      <?PHP } ?>
    </td>
    <td align="right">
      <?PHP
        if (isset($global_status)) {
          echo $global_status;
        }
      ?>
    </td>
  </tr>
</table>
</form>
</div>
</body>
</html>

<?PHP
/***************************************************************************
 * $Log: taskbar.php,v $
 * Revision 1.6  2003/12/07 19:21:16  willuhn
 * @C language
 * @B
 *
 * Revision 1.5  2003/11/18 02:00:51  willuhn
 * *** empty log message ***
 *
 * Revision 1.4  2003/10/07 20:35:18  willuhn
 * @B fixed bug (sourceforge bug id: 815180)
 *
 * Revision 1.3  2003/09/27 18:28:16  willuhn
 * @B fixed typo in taskbar.php
 *
 * Revision 1.2  2003/09/27 18:23:45  willuhn
 * *** empty log message ***
 *
 * Revision 1.1.1.1  2003/07/28 19:22:29  willuhn
 * reimport
 *
 * Revision 1.21  2002/11/07 22:57:21  willuhn
 * @B division by zero in "project roadmap" plugin
 * @B some calculation errors in plugins "task hotlist" and "project roadmap" fixed
 * @N renamed page "password" into "preferences"
 * @N user is now able to change his language settings
 * @N added some constants for task properties
 *
 * Revision 1.20  2002/06/26 14:14:29  willuhn
 * @N added form "query task by id"
 *
 * Revision 1.19  2002/05/03 11:31:53  willuhn
 * @B session handling
 *
 * Revision 1.18  2002/05/02 19:51:32  willuhn
 * @N task->isAvailable() checks if task is "in progress" or "request"
 * @N link to parent task in taskdetails
 * @N link to all child tasks in taskdetails
 * @B in task hotlist
 *
 * Revision 1.17  2002/04/17 22:26:46  willuhn
 * @N child tasks can only set to "in progress", if parent task is set to "done"
 *
 * Revision 1.16  2002/04/17 19:54:43  willuhn
 * @B a lot of fixes for "register_globals=off"
 *
 * Revision 1.15  2002/03/29 01:50:24  willuhn
 * @N merged template bill and joblist
 * @N performance speedups by caching frequently used values (rights,prios,types...)
 *
 * Revision 1.14  2002/02/09 19:38:27  willuhn
 * @N added CVS log
 * @N added french language file
 *
 *
 ***************************************************************************/
?>
